lib/semian/sysv_shared_memory.rb

```diff
@@ -0,0 +1,75 @@
+module Semian
+  module SysVSharedMemory #:nodoc:
+    @type_size = {}
+    def self.sizeof(type)
+      size = (@type_size[type.to_sym] ||= (respond_to?(:_sizeof) ? _sizeof(type.to_sym) : 0))
+      raise TypeError.new("#{type} is not a valid C type") if size <= 0
+      size
+    end
+
+    def self.included(base)
+      def base.do_with_sync(*names)
```

Comment:
I see what you're saying, but that also doesn't work in protecting lets say.. `Semian::SysV::Integer#value`. If the `#value` is not protected in C code, and it directly replaces the `SysV::Integer`, there is no entry point to override and replace with `synchronize { ... }` except through what `do_with_sync` does. If you do protect in C, well, that goes back to the problem of making 2x the functions to call `rb_ensure(callback, arg1, ensure_fn, arg2)`